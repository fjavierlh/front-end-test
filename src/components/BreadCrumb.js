import React from 'react';
import { Link, useLocation } from 'react-router-dom';
import styled from 'styled-components';

export const BreadCrumb = () => {
  const location = useLocation();
  const state = location.state;
  console.log(`state`, state);

  if (state === null) {
    return (
      <BreadCrumbStyled>
        <span>
          <b>Home</b>
        </span>
      </BreadCrumbStyled>
    );
  }

  return (
    <BreadCrumbStyled>
      <Link to="/">
        <span>Home</span>
      </Link>
      <span className="separator">â¯ˆ</span>
      <span>
        <b>{state.model}</b>
      </span>
    </BreadCrumbStyled>
  );
};

const BreadCrumbStyled = styled.div`
  & > .separator {
    margin: 0 5px;
    color: #000;
  }
  & > * {
    text-decoration: none;
    display: inline-block;
    color: #666;
  }
`;
