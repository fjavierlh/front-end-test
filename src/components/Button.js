import React from 'react';

export const Button = ({ text, children, disabled }) => {
  return <button disabled={disabled}>{text ? text : children}</button>;
};
