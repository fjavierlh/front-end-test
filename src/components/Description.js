import React from 'react';
import styled from 'styled-components';

export const Description = ({ product }) => {
  const {
    brand,
    model,
    price,
    cpu,
    ram,
    os,
    battery,
    primaryCamera,
    secondaryCamera,
    displayResolution,
    dimensions,
    weight,
    displayType,
    displaySize,
    chipset,
    externalMemory,
    internalMemory,
    nfc,
    usb,
  } = { ...product };
  return (
    <DescriptionStyled>
      <div className="description">
        <dl className="description-left">
          {brand && (
            <>
              <dt>Brand</dt>
              <dd>{brand}</dd>
            </>
          )}
          {model && (
            <>
              <dt>Model</dt>
              <dd>{model}</dd>
            </>
          )}
          {price && (
            <>
              <dt>Price</dt>
              <dd>{price}â‚¬</dd>
            </>
          )}
          {cpu && (
            <>
              <dt>CPU</dt>
              <dd>{cpu}</dd>
            </>
          )}
          {ram && (
            <>
              <dt>RAM</dt>
              <dd>{ram}</dd>
            </>
          )}
          {os && (
            <>
              <dt>OS</dt>
              <dd>{os}</dd>
            </>
          )}
          {displayResolution && (
            <>
              <dt>Display Resolution</dt>
              <dd>{displayResolution}</dd>
            </>
          )}
          {battery && (
            <>
              <dt>Battery</dt>
              <dd>{battery}</dd>
            </>
          )}
          {dimensions && (
            <>
              <dt>Dimensions</dt>
              <dd>{dimensions}</dd>
            </>
          )}
          {weight && (
            <>
              <dt>Weight</dt>
              <dd>{weight}</dd>
            </>
          )}
        </dl>
        <dl className="description-left">
          {(primaryCamera || secondaryCamera) && (
            <>
              <dt>Cameras</dt>
              <dd>
                {primaryCamera && (
                  <>
                    <dt>Primary Camera</dt>
                    <dd>{primaryCamera}</dd>
                  </>
                )}
                {secondaryCamera && (
                  <>
                    <dt>Secondary Camera</dt>
                    <dd>{secondaryCamera}</dd>
                  </>
                )}
              </dd>
            </>
          )}

          {displayType && (
            <>
              <dt>Display Type</dt>
              <dd>{displayType}</dd>
            </>
          )}

          {displaySize && (
            <>
              <dt>Display Size</dt>
              <dd> {displaySize}</dd>
            </>
          )}

          {chipset && (
            <>
              <dt>Chipset</dt>
              <dd>{chipset}</dd>
            </>
          )}
          {externalMemory && (
            <>
              <dt>External Memory</dt>
              <dd>{externalMemory}</dd>
            </>
          )}
          {internalMemory && (
            <>
              <dt>Internal Memory</dt>
              <dd>{internalMemory}</dd>
            </>
          )}

          {nfc && (
            <>
              <dt>NFC</dt>
              <dd>{nfc}</dd>
            </>
          )}
          {usb && (
            <>
              <dt>USB</dt>
              <dd>{usb}</dd>
            </>
          )}
        </dl>
      </div>
    </DescriptionStyled>
  );
};

const DescriptionStyled = styled.article`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;

  dl {
    margin: 0;
  }

  dt {
    text-transform: uppercase;
    font-size: 0.9em;
    font-weight: bold;
    margin-top: 0.5rem;
    margin-top: 0.8rem;
  }

  dd > dt {
    font-size: 0.8em;
    margin-left: 2em;
    margin-bottom: 0.5rem;
  }
  dd > dd {
    font-size: 0.8em;
    margin-left: 2em;
  }

  dd {
    font-size: 0.9rem;
    margin: 0;
    font-weight: lighter;
  }

  & > .description {
    display: flex;
    gap: 1.5em;

    justify-content: space-evenly;
  }
`;
