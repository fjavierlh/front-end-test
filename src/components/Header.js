import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

export const Header = ({ logoUrl, title, cartCount }) => {
  return (
    <HeaderStyled>
      <Link to="/" className="brand">
        {logoUrl ? (
          <img className="logo" src={logoUrl} alt="logo"></img>
        ) : (
          <h4>{title}</h4>
        )}
      </Link>
      <div className="cart">
        <p>Items in your cart</p>
        <div className="cart-count">{cartCount}</div>
      </div>
    </HeaderStyled>
  );
};

const HeaderStyled = styled.header`
  display: flex;
  justify-content: space-between;
  align-items: center;
  & > .logo {
    width: 200px;
    max-width: 200px;
  }
  & > .brand {
    text-decoration: none;
    color: #333;
  }
  & > .cart {
    display: flex;
    flex-wrap: wrap;
    gap: 0.5em;
    text-transform: uppercase;
    align-items: center;
    font-size: 0.9em;
    font-weight: bold;
  }
  & > .cart > .cart-count {
    display: flex;
    margin-right: 1rem;
    justify-content: center;
    align-items: center;
    background-color: #333;
    color: #fff;
    height: 1.5rem;
    width: 1.5rem;
    font-weight: bold;
    font-size: 0.8rem;
    border-radius: 100%;
  }
`;

export default Header;
