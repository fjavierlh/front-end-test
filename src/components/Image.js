import React from 'react';

export const Image = ({ imgUrl, altText, lazyLoading }) => {
  return <img src={imgUrl} alt={altText} loading={lazyLoading ? 'lazy' : ''} />;
};
