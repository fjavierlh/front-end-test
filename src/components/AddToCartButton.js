import React from 'react';
import { Button } from './Button';

export const AddToCartButton = ({ userOptions, text, children }) => {
  const disabled = userOptions
    ? !userOptions.colorCode || !userOptions.storageCode
    : true;

  return <Button text={text} children={children} disabled={disabled} />;
};
