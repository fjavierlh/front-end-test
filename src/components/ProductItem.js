import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { Image } from './Image';

const ProductItem = ({ id, imgUrl, brand, model, price }) => {
  return (
    <ProductItemStyled>
      <Link
        to={`/product/${model}`.toLowerCase().replaceAll(' ', '-')}
        className="product-link"
        key={id}
        state={{ id, imgUrl, brand, model }}
      >
        <Image
          className="img-item"
          imgUrl={imgUrl}
          altText={`${brand}-${model}`}
          lazyLoading={true}
        />
        <div className="product-info">
          <p>
            <b>{brand}</b> {model}
          </p>
          <span className="price-text">
            {price ? <b>{price}â‚¬</b> : <small>Price to be consulted</small>}
          </span>
        </div>
      </Link>
    </ProductItemStyled>
  );
};

const ProductItemStyled = styled.article`
  flex: 1 0 21%;
  min-height: 180px;
  & > .product-link {
    flex: 1 0 21%;
    flex-basis: 100%;
    text-decoration: none;
    color: #333;

    & > .product-info {
      display: flex;
      flex-direction: column;
      justify-content: flex-end;
    }
  }
  text-align: center;
  background-color: #fff;
  padding: 2rem 0 2rem 0;

  & > .img-item {
    display: block;
    max-height: 180px;
    min-height: 180px;

    margin: 0 2em;
  }

  & > .product-info {
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
  }
`;

export default ProductItem;
