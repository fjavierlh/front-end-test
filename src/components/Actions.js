import React from 'react';
import styled from 'styled-components';
import { AddToCartButton } from './AddToCartButton';

export const Actions = ({
  id,
  onSubmit,
  onChangeColor,
  onChangeStorage,
  options,
  userOptions,
}) => {
  const { colors, storages } = options;
  return (
    <ActionsStyled>
      <form onSubmit={onSubmit}>
        <select onChange={onChangeColor}>
          <option key={`${id}-colors`} disabled selected>
            Colors
          </option>
          {colors &&
            colors.map(({ name, code }) => (
              <option key={`${id}-${code}`} value={code}>
                {name}
              </option>
            ))}
        </select>
        <select onChange={onChangeStorage}>
          <option key={`${id}-storages`} disabled selected>
            Storages
          </option>
          {storages &&
            storages.map(({ name, code }) => (
              <option key={`${id}-${code}`} value={code}>
                {name}
              </option>
            ))}
        </select>
        <AddToCartButton userOptions={userOptions} text="Add to cart" />
      </form>
    </ActionsStyled>
  );
};

const ActionsStyled = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;

  & > h3 {
    flex-basis: 100%;
  }
  & > form {
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin-top: 1rem;
  }
  & > form > select {
    font-size: 1.2rem;
    display: block;
    width: 320px;
    border: 1px solid #ccc;
    border-radius: 5px;
    background-color: #fff;
    margin: 0 auto 1rem auto;
    padding: 0.5rem;
  }
`;
