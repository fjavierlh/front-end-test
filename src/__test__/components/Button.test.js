import { render, screen } from '@testing-library/react';
import { Button } from '../../components/Button';

describe('Button', () => {
  test('should render Button', () => {
    const text = 'Click me!';
    render(<Button text={text} />);
    expect(screen.getByText(text)).toBeInTheDocument();
  });

  test('should render Button with default text', () => {
    const defaultText = 'Click me!';
    render(<Button />);
    expect(screen.getByText(defaultText)).toBeInTheDocument();
  });

  test('should render Button disable', () => {
    const defaultText = 'Click me!';
    render(<Button disabled={true} />);
    expect(screen.getByText(defaultText)).toBeDisabled();
  });

  test('should render Button enable', () => {
    const defaultText = 'Click me!';
    render(<Button disabled={false} />);
    expect(screen.getByText(defaultText)).not.toBeDisabled();
  });

  test('should match snapshot button when is Button enable', () => {
    const { container } = render(<Button disabled={true} />);
    expect(container).toMatchSnapshot();
  });
});
