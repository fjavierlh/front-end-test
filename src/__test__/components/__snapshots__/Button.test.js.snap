// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot button when is Button enable 1`] = `
<div>
  <button
    class="sc-bdvvtL aFQfL"
    disabled=""
  >
    Click me!
  </button>
</div>
`;
